1_data_prep/13_tree_crops.py

```diff
@@ -0,0 +1,353 @@
+            species_dir_label = 'unlabelled' if final_species is None else 'labelled'
+
+            if not should_save_crop:
+                continue
+
+            ys, xs = np.where(mask_ids == tree_mask_value) # gather all pixels of current tree
+            if ys.size == 0 or xs.size == 0:
+                continue  # skip if no pixels found
+
+            y0, y1, x0, x1 = ys.min(), ys.max(), xs.min(), xs.max() # current tree img bounds
+
+            # check for 'bad' crops
+            # currently looks for low res images (h or w < 200 px), 
+            # or crops near the edge of the images liable for distortion (any bbox edge <200px away from full img edge)
+            # TODO: also try laplacian variance and/or tenengrad score, as well as checking for low contrast
+
+            # low res img
+            # typically trees too far off in the distance
+            if y1 - y0 < IMAGE_RES_CONSTRAINT or x1 - x0 < IMAGE_RES_CONSTRAINT:
+                continue
+
+            # check if image within safe radius to limit distortion effects
+            crop_cx, crop_cy = (x0 + x1) / 2, (y0 + y1) / 2
+            diff = np.array([img_cx - crop_cx, img_cy - crop_cy])
+            if diff @ diff > safe_radius ** 2: # similar to euclidean norm but without sqrt -> faster
+                continue
+
+            # cropped near edge
+            # near edge may not capture whole tree
+            # radial detection above should capture most of these but this exists as a failsafe
+            if img.height - y1 < IMAGE_RES_CONSTRAINT or img.width - x1 < IMAGE_RES_CONSTRAINT \
+            or y0 < IMAGE_RES_CONSTRAINT or x0 < IMAGE_RES_CONSTRAINT:
+                continue
+
+            # crop image to bbox plus padding
+            # cropping in PIL is left, top, right, bottom
+            bbox_pad_width, bbox_pad_height = (x1 - x0)  * BBOX_PADDING_RATIO, (y1 - y0) * BBOX_PADDING_RATIO
+            bbox = (x0 - bbox_pad_width, y0 - bbox_pad_height, x1 + bbox_pad_width, y1 + bbox_pad_height)
+            cropped_img = img.crop(bbox)
+
+            fp = Path(
+                output_path,                        # base dir to the training data folder
+                species_dir_label,                  # separate trees with species labels or no species labels
+                src_info['dset_name'],              # original dataset name (src folder name of images/masks)
+                f"treeID{tree_unique_id}"           # tree id from mask
+            ).resolve()
+            os.makedirs(fp, exist_ok=True)
+
+            # generate random 10-digit ID for the image
+            image_id = str(uuid.uuid4().int)[:10]   # take first 10 digits of a UUID-derived int
+            crop_path = fp / f"{image_id}.png"
+
+            cropped_img.save(crop_path) # save cropped img
+            mapping_stats['saved_crops'] += 1
+
+            # Build metadata record with dynamic columns based on mapping level
+            record = {
```

Comment:
I wonder if it would be helpful to include the `img_file_path` here for debugging?